cpp/src/io/parquet/page_enc.cu

```diff
@@ -1153,45 +1150,60 @@ inline __device__ uint8_t *cpw_put_fldh(uint8_t *p, int f, int cur, int t)
   }
 }
 
-#define CPW_BEGIN_STRUCT(hdr_start) \
-  {                                 \
-    uint8_t *p  = hdr_start;        \
-    int cur_fld = 0;
-
-#define CPW_FLD_STRUCT_BEGIN(f)                         \
-  p       = cpw_put_fldh(p, f, cur_fld, ST_FLD_STRUCT); \
-  cur_fld = 0;
-
-#define CPW_FLD_STRUCT_END(f) \
-  *p++    = 0;                \
-  cur_fld = f;
-
-#define CPW_FLD_INT32(f, v)                          \
-  p       = cpw_put_fldh(p, f, cur_fld, ST_FLD_I32); \
-  p       = cpw_put_int32(p, v);                     \
-  cur_fld = f;
-
-#define CPW_FLD_INT64(f, v)                          \
-  p       = cpw_put_fldh(p, f, cur_fld, ST_FLD_I64); \
-  p       = cpw_put_int64(p, v);                     \
-  cur_fld = f;
-
-#define CPW_FLD_BINARY(f, v, l)                   \
-  p = cpw_put_fldh(p, f, cur_fld, ST_FLD_BINARY); \
-  p = cpw_put_uint32(p, l);                       \
-  memcpy(p, v, l);                                \
-  p += l;                                         \
-  cur_fld = f;
-
-#define CPW_END_STRUCT(hdr_end) \
-  *p++    = 0;                  \
-  hdr_end = p;                  \
+struct cpw {
```

Comment:
We can take this opportunity to also improve the naming here.
I'd also make this a class, since p no one needs to access the data members directly.
```suggestion
class compact_protocol_writer {
```

Reply:
Changing it to `header_encoder`.